### PR TITLE
Public Backend API

### LINKED ISSUE #604: API for Additional Backends
Many users have a preferred web or cloud-based interface for running Stable Diffusion besides the 🤗 diffusers based implementation Dream Textures ships with. Dream Textures currently supports swapping out the backend to support both diffusers and DreamStudio, but new backends must be maintained here. This proposes a new architecture that would allow other addons to add backends to Dream Textures.

## Detailed Design
Each backend would be represented as its own Blender addon. This addon would only need to implement a limited set of functionality to provide a backend for Dream Textures.

### Backend Registration
In an addon's `register` function, it should register a backend class with blender.

```python
def register():
    bpy.utils.register_class(Automatic1111)
```

Dream Textures can then look up all registered backends from the `bpy.types` module.

### Backend Interface
Backends would themselves be `bpy.types.PropertyGroup` subclasses. This would allow them to declare any additional properties or optimizations they support on their class.

A backend would implement methods like the following:

```python
def list_models(self) -> list[Model]: ...
```
Returns the list of models installed for this backend.

Model import/downloading would be handled by each individual backend's addon in its own preferences window.

```python
def supported_features(self) -> set[SupportedFeature]: ...
```
Can be used to enable certain aspects of the main interface.

Possible values could be:
* `SEAMLESS`
* `NEGATIVE_PROMPT`
* `CONTROL_NET`

```python
def draw_prompt(self, layout, context): ...
def draw_speed_optimizations(self, layout, context): ...
def draw_memory_optimizations(self, layout, context): ...
def draw_extra(self, layout, context): ...
```
The backend should draw its own properties in one of these sections.

```python
def prompt_to_image(self, dream_prompt, on_response, on_complete, on_exception): ...
def image_to_image(self, dream_prompt, on_response, on_complete, on_exception): ...
def depth_to_image(self, dream_prompt, on_response, on_complete, on_exception): ...
def inpaint(self, dream_prompt, on_response, on_complete, on_exception): ...
def outpaint(self, dream_prompt, on_response, on_complete, on_exception): ...
def control_net(self, dream_prompt, on_response, on_complete, on_exception): ...
```
If any of these methods are declared, it is assumed that the task is supported by the backend. Otherwise, it will be reported to the user that the backend cannot be used for that task.

### Dream Textures API
To make imports simple for backend addons, a separate `api` folder could contain all public classes necessary for creating backends.

```python
from dream_textures.api import Model, SupportedFeature, DreamPrompt
```

This module could also contain a better public API for calling Dream Textures from scripts running via Blender's background mode.

## Alternatives Considered
The most popular backends could be added natively to Dream Textures. However, supporting every backend users may want would be impossible to keep up to date within this single repo with only a few contributors.

### PR DESCRIPTION
**Pull request by carson-katri**
Closes #604 

The API is slightly different than it that issue. You can see an example of a backend in `diffusers_backend.py`, which is the default Dream Textures backend.

```python
class MyCustomBackend(Backend):
    name = "My Backend Name"
    description = "A short description of this backend"

    custom_optimization: BoolProperty(name="My Custom Optimization")

    def list_models(self, context) -> List[Model]:
        return [...]

    def list_schedulers(self, context) -> List[str]:
        return [...]

    def generate(self, task: Task, model: Model, prompt: Prompt, size: Tuple[int, int] | None, seed: int, steps: int, guidance_scale: float, scheduler: str, seamless_axes: SeamlessAxes, step_preview_mode: StepPreviewMode, iterations: int, step_callback: StepCallback, callback: Callback):
        match task:
            case PromptToImage():
                ...
            case ImageToImage(image=image, strength=strength, fit=fit):
                ...
            case Inpaint(image=image, fit=fit, strength=strength, mask_source=mask_source, mask_prompt=mask_prompt, confidence=confidence):
                ...
            case DepthToImage(depth=depth, image=image, strength=strength):
                ...
            case Outpaint(image=image, origin=origin):
                ...
            case _:
                raise NotImplementedError()

    def draw_speed_optimizations(self, layout, context):
        layout.prop(self, "custom_optimization")
```

A backend could be as simple as a single file Blender add-on, so I'm considering adding a folder to this repo to contain scripts for additional backends that would be maintained by different community members. This could contain backends for sources like Automatic1111, DreamStudio, InvokeAI, StableHorde, etc.

### CODE REVIEW

**Comment by NullSenseStudio** [comment]
Looks like you've accidentally pushed the packaged CoreML backend in https://github.com/carson-katri/dream-textures/pull/654/commits/47118a07267bdb8b20b20d3a09e1ad8f4be1bda4

**Comment by carson-katri** [comment]
Should be fixed now.

**Comment by carson-katri** [comment]
Basic image generation is working. Here's a list of features that need to be tested/updated:
- [x] Prompt to image
- [ ] Image to image
- [ ] Inpainting
- [ ] Depth to image
- [ ] ControlNet
- [ ] Texture Projection
- [ ] Nodes
- [ ] Render Pass
- [ ] Upscaling

**Comment by Made-0f-Ice** [comment]
I just tried to test this tool, I have "automatic1111" server running on 127.0.0.1:7860 (working via web-interface), but on the "Dream" panel in Blender I see only "Huggingface Diffusers" backend and no available models… (I'm on Debian Linux)
Does this approch yet to implement or I've made some error configuring?
Thank you in advance. I'm not hurry or so, I'm just interested in this feature.

**Comment by carson-katri** [comment]
This PR doesn’t add an Automatic1111 backend, it just makes it possible for one to be created in the future.

**Comment by macronomicus** [comment]
So you dont plan to use ckpt/safetensors files, but rather allow for other backend? Is there performance loss with those types? Or just too much refactoring? I've read a few posts & was curious. 

I had dreamtexture installed on last pc but have yet to dive back in on new setup, im glad you're still progressing & its looking great, blender is by far my favorite app. It'd be nice to use our model files from automatic1111, but regardless Dreamtextures stands on its own without it. I am fine with the two being separate, its simple enough to jump back & forth between them with stuff. But you're showing such great & deeper integrations its very tempting!

I imagine blender official will get in on the ai stuff too, its all still fast moving but at some point you might get some official specialized blender api stuff to work with too. Im keeping an eye out for when they start talking more ai.

**Comment by carson-katri** [comment]
@macronomicus Diffusers added a new [`from_single_file`](https://huggingface.co/docs/diffusers/main/en/api/pipelines/stable_diffusion/text2img#diffusers.StableDiffusionPipeline.from_single_file) loader, which can load `.ckpt`/`safetensor` files. With that we may be able to easily support separate lookup directories for single-file models.

**Comment by NullSenseStudio** [comment]
I'll take a crack at adding `from_single_file` since I'm already handling related code in #703

**Review by NullSenseStudio — CHANGES_REQUESTED** [review]

File: generator_process/models/optimizations.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        except: pass
+
+        try:
+            if self.vae_tiling != "off":
+                if not isinstance(pipeline.decode_latents, functools.partial):
+                    pipeline.decode_latents = functools.partial(tiled_decode_latents.__get__(pipeline), pre_patch=pipeline.decode_latents)
+                pipeline.decode_latents.keywords['optimizations'] = self
+            elif self.vae_tiling == "off" and isinstance(pipeline.decode_latents, functools.partial):
+                pipeline.decode_latents = pipeline.decode_latents.keywords["pre_patch"]
+        except: pass
```
[the remark this answers is not in the document]
Comment (carson-katri) [reply generator_process/models/optimizations.py:163]: @NullSenseStudio Feel free to take it

File: property_groups/dream_prompt_validation.py [hunk 2]
```diff
@@ -1,6 +1,5 @@
 from ..preferences import StableDiffusionPreferences, _template_model_download_progress, InstallModel
-from ..generator_process.models import Pipeline, FixItError
-from ..generator_process.actions.huggingface_hub import ModelType
+from ..generator_process.models import ModelType
 from ..preferences import OpenURL
 
 def validate(self, context, task: ModelType | None = None) -> bool:
```
[the remark this answers is not in the document]
Comment (carson-katri) [reply property_groups/dream_prompt_validation.py]: Should only be in the backend now, unless its something that the backend cannot validate (like selected faces for projection).

**Comment by NullSenseStudio** [comment]
Took a bit more than I thought to fix VAE tiling. It's a bit of a mess with how it's juggling the image between torch and numpy, but it's functional for now. I'll get that in better order soon since I have some changes planned for VAE tiling.

Outpainting sort of works, but the outpainted regions remain black. Seems to be caused by changes made to the inpaint pipeline. You can color a portion of an image then mark that area for inpainting, and it will still resemble that color after inpainting now.

**Comment by carson-katri** [comment]
Could the `strength` value be incorrect, so it takes some of the black into account?

**Comment by NullSenseStudio** [comment]
That was it!

**Review by NullSenseStudio — APPROVED** [review]

**Comment by TimWolfram** [comment]
@carson-katri 

> This PR doesn’t add an Automatic1111 backend, it just makes it possible for one to be created in the future.
> 
> Contributions for new backends would be welcome once this is merged!

Has nobody implemented an A1111 backend yet? I can't find any mentions of it existing but considering it's been a year now since the backend API and A1111 is the most popular way to use stable diffusion, I'd assume someone would have implemented it by now. If not, I can try to implement it myself but I thought I might ask here before wasting my time.

**Comment by carson-katri** [comment]
@TimWolfram As far as I know, no one has built an A1111 backend yet. But I'd be happy to answer any questions you have if you build one.

**Pull request merged**